Create new node tool

Divinare opened the pull request:
Closes #604

BE PR: https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/31

* moved isSelected to nodeMarker props to solve problem where at node/new node was not being selected (now it is because we can say node is selected by checking if url has subSites.newNode)

* moved isSelected to nodeMarker props to solve problem where at node/new node was not being selected (now it is because we can say node is selected by checking if url has subSites.newNode)

* Centering map is now EditNodeLayer's responsibility --> map is centered after new node is created

* Fixed bug of node change being called multiple times

* "Network should not refresh when dragging node"
--> this is how our vectorGridLayer works at the moment. We have "// TODO: find a better way to achieve this". Didn't make a new card to fix this because lag issue is not fixed, it was because our node was changed 9 times due to bugs in NodeCoordinatesView

TODO later:
After node is created, redirect to node/id: CARD: #666 

How to test:

1. Create a new node
1. After save, backend console should log new node id
1. Paste new node id to url: http://localhost:3000/node/{nodeId}

kronken commented:
Works rather nicely.
Bugs:
- Map is moved anytime any node is moved
- Last node is missing on all routepaths in route path list view, added review comment what causes it.
- Route paths are still visible on map, even after starting to add new nodes
- Saving the new node twice, causes the BE to create to identical nodes. I know this will be fixed by #666, but it will cause a rather bad bug in master. Should we just disable saving button after once saved?

Divinare commented on src/components/map/layers/edit/EditNodeLayer.tsx:
```diff
@@ -49,6 +48,16 @@ class EditNodeLayer extends Component<IEditNodeLayerProps> {
+        const node = this.props.nodeStore!.node;
+        if (!node) {
+            this.removeOldLinks();
+        }
```
This func no longer exists

Divinare commented on src/components/map/layers/edit/EditNodeLayer.tsx:
```diff
@@ -57,14 +66,20 @@ class EditNodeLayer extends Component<IEditNodeLayerProps> {
     }
 
+    private centerNode = (node: INode) => {
+        this.props.mapStore!.setCoordinates(node.coordinates);
```
Now it has two lines of code. I think its fine.

Divinare commented on src/components/map/toolbar/ToolbarCommonButtons.tsx:
```diff
@@ -1,8 +1,8 @@
 import React from 'react';
-import { FiPrinter, FiExternalLink } from 'react-icons/fi';
+import { FiPrinter, FiExternalLink, FiPlus } from 'react-icons/fi';
```
I think we might need graphical designer to create us some icons. It can be changed later.

Divinare commented on src/stores/toolbarStore.ts:
```diff
@@ -13,14 +13,12 @@ const TOOL_LIST = [
-    return Object.assign({ [c.toolType]:c }, acc);
-});
+const TOOLS = {};
+TOOL_LIST.map((tool: BaseTool) => TOOLS[tool.toolType] = tool);
```
The code before didn't quite work when selecting tools. And its now a lot easier to read

Divinare commented:
Good comments. Review fixes done.

* Now EditNodeLayer is only rendered if node exists at nodeStore

-Map is moved anytime any node is moved
   -> Now map is centered when node changes from null -> node
-Last node is missing on all routepaths in route path list view, added review comment what causes it.
   -> Fixed, thx for pointing it out
-Route paths are still visible on map, even after starting to add new nodes
   -> Fixed, now it clears routePaths after create node tool is used
-Saving the new node twice, causes the BE to create to identical nodes. I know this will be fixed by #666, but it will cause a rather bad bug in master. Should we just disable saving button after once saved?
   -> Now it gets temporarily redirected to /home page. Added TODO there

Divinare commented:
Also fixed one more bug.

Can be reviewed again.

Divinare commented:
Could we use another cursor when pointing where to create node. Currently there is the 'move hand' cursor, which doesn't really mean pointing or creating.
--> Now there is a crosshair cursor when creating new node

Creating a new node breaks undo/redo for node geometries. This will only happen when creating a new node, but after creating a new node, it will also happen for old nodes.
--> Fixed now all undo/redo of links, nodes, routePaths

Unable to save crossroad. Save fails, BE says that it tries to sync stop for the node. It should probably not throw error if there is no localStop or joreStop. This just means that there is nothing to be done.
--> Fixed in BE pr (remember to pull in backend)

Save button is disabled for new crossroad
--> is an old bug, card: #721
